tests/core/test_kontainer_drivers.py

```diff
@@ -94,9 +95,73 @@ def test_enabling_driver_exposes_schema(admin_mc, remove_resource):
     kd.active = False
     admin_mc.client.update_by_id_kontainerDriver(kd.id, kd)
 
+    admin_mc.client.delete(kd)
+    verify_driver_not_in_types(admin_mc.client, kd)
+
+
+@pytest.mark.nonparallel
+def test_create_duplicate_driver_conflict(admin_mc, remove_resource):
+    """ Test if adding a driver with a pre-existing driver's URL
+    returns a conflict error"""
+    kd = admin_mc.client.create_kontainerDriver(
+        createDynamicSchema=True,
+        active=True,
+        url=DRIVER_URL
+    )
+    remove_resource(kd)
+
+    kd = wait_for_condition('Active', 'True', admin_mc.client, kd, timeout=90)
+    verify_driver_in_types(admin_mc.client, kd)
+
+    try:
+        kd2 = admin_mc.client.create_kontainerDriver(
+            createDynamicSchema=True,
+            active=True,
+            url=DRIVER_URL
+        )
+        remove_resource(kd2)
+    except ApiError as e:
+        assert e.error.status == 409
+        assert "Driver URL already in use:" in e.error.message
+
+    admin_mc.client.delete(kd)
     verify_driver_not_in_types(admin_mc.client, kd)
 
 
+@pytest.mark.nonparallel
+def test_update_duplicate_driver_conflict(admin_mc, remove_resource):
+    """ Test if updating a driver's URL to a pre-existing driver's URL
+    returns a conflict error"""
+    kd1 = admin_mc.client.create_kontainerDriver(
+        createDynamicSchema=True,
+        active=True,
+        url=DRIVER_URL
+    )
+    remove_resource(kd1)
+
+    kd1 = wait_for_condition('Active', 'True', admin_mc.client, kd1,
+                             timeout=90)
+
+    kd2 = admin_mc.client.create_kontainerDriver(
+        createDynamicSchema=True,
+        active=True,
+        url=DRIVER_URL + "2"
+    )
+    remove_resource(kd2)
+
+    kd2.url = DRIVER_URL
+    try:
+        admin_mc.client.update_by_id_kontainerDriver(kd2.id, kd2)
+    except ApiError as e:
+        assert e.error.status == 409
+        assert "Driver URL already in use:" in e.error.message
+
+    admin_mc.client.delete(kd1)
+    verify_driver_not_in_types(admin_mc.client, kd1)
+
+    admin_mc.client.delete(kd2)
```

Comment:
Fix remove resource so it properly cleans up kd2 rather than this explicit delete.